### PR TITLE
JAPI-424 Provide compression algorithm

### PR DESCRIPTION
**Pull request by jpmcmu**
Signed-off-by: James McMullan James.McMullan@lexisnexis.com

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [X] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).

## Checklist:
- [X] I have created a corresponding JIRA ticket for this submission
- [X] My code follows the code style of this project.
  - [ ] I have applied the Eclipse code-format template provided.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [X] I have read the HPCC Systems CONTRIBUTORS document (https://github.com/hpcc-systems/HPCC-Platform/wiki/Guide-for-contributors).
- [X] The change has been fully tested:
  - [ ] I have performed unit tests to cover my changes.
  - [ ] I have performed system test and covered possible regressions and side effects.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [X] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.

### CODE REVIEW

**Comment by jpmcmu** [comment]
@rpastrana still tracking down a few bugs

**Review by rpastrana — COMMENTED** [review]
@jpmcmu how much of this comes from "FastLZ - Byte-aligned LZ77 compression library"?
Can that logic be java imported, or submoduled from their github repo? then we can build a wrapper around it?

**Comment by jpmcmu** [comment]
@rpastrana The FastLZ code is in C. There are a few ports out there but they don't have licenses attached to them. I added the license to the code I ported over to remind me to check if it needs to be included, since I used the C code as a starting point.

**Comment by jpmcmu** [comment]
@rpastrana please review

File: commons-hpcc/src/main/java/org/hpccsystems/commons/fastlz/FastLZ.java [hunk 1]
```diff
@@ -0,0 +1,833 @@
+/*******************************************************************************
```
Comment (rpastrana) [inline commons-hpcc/src/main/java/org/hpccsystems/commons/fastlz/FastLZ.java:1]: does it make sense to name this fastLZ4j ?

File: commons-hpcc/src/main/java/org/hpccsystems/commons/fastlz/FastLZ.java [hunk 2]
```diff
@@ -0,0 +1,833 @@
...
+        // unknown level, trigger error
+        return 0;
+    }
+
+    public static boolean decompress_streaming(DecompressionState state, byte[] input, int inputAvailableLength, byte[] output) throws Exception
+    {
+        // magic identifier for compression level
+        if (state.compressionLevel < 0)
+        {
+            state.compressionLevel = (input[0] >> 5) + 1;
```
Comment (rpastrana) [inline commons-hpcc/src/main/java/org/hpccsystems/commons/fastlz/FastLZ.java:784]: we might need to guard against empty input array

File: commons-hpcc/src/main/java/org/hpccsystems/commons/fastlz/FastLZ.java [hunk 3]
```diff
@@ -0,0 +1,833 @@
...
+        return true;
+    }
+
+    public static int compress(byte[] input, byte[] output)
+    {
+        // for short block, choose fastlz1
+        if (input.length < 65536)
+            return fastlz1_compress(input, output);
+
+        // else...
```
Comment (rpastrana) [inline commons-hpcc/src/main/java/org/hpccsystems/commons/fastlz/FastLZ.java:763]: might as well code in the "else"

File: commons-hpcc/src/main/java/org/hpccsystems/commons/fastlz/FastLZ.java [hunk 4]
```diff
@@ -0,0 +1,833 @@
...
+            ctrl = inputBuffer[ip] & 0xFF;
+            ip++;
+        }
+
+        return op;
+    }
+
+    public static class DecompressionState
+    {
+        private int ip = -1;
```
Comment (rpastrana) [inline commons-hpcc/src/main/java/org/hpccsystems/commons/fastlz/FastLZ.java:602]: can we use a self describing name for these vars?

**Review by rpastrana — CHANGES_REQUESTED** [review]
@jpmcmu added a few questions. 
Also, not all methods need a javadoc comment, but let's try to javadoc comment the 2 classes and perhaps the major methods, and perhaps a highlevel package level write up like this one:
HPCC4J\wsclient\src\main\java\org\hpccsystems\ws\client\package-info.java

**Review by rpastrana — APPROVED** [review]

**Pull request merged**